Parser/action_helpers.c

```diff
@@ -1612,19 +1612,44 @@ _build_concatenated_bytes(Parser *p, asdl_expr_seq *strings, int lineno,
     Py_ssize_t len = asdl_seq_LEN(strings);
     assert(len > 0);
 
-    PyObject* res = Py_GetConstant(Py_CONSTANT_EMPTY_BYTES);
-
     /* Bytes literals never get a kind, but just for consistency
         since they are represented as Constant nodes, we'll mirror
         the same behavior as unicode strings for determining the
         kind. */
-    PyObject* kind = asdl_seq_GET(strings, 0)->v.Constant.kind;
+    PyObject *kind = asdl_seq_GET(strings, 0)->v.Constant.kind;
+
+    Py_ssize_t total = 0;
+    for (Py_ssize_t i = 0; i < len; i++) {
+        expr_ty elem = asdl_seq_GET(strings, i);
+        Py_ssize_t part = PyBytes_GET_SIZE(elem->v.Constant.value);
+        if (part > PY_SSIZE_T_MAX - total) {
+            PyErr_NoMemory();
+            return NULL;
+        }
+        total += part;
+    }
+
+    PyBytesWriter *writer = PyBytesWriter_Create(total);
+    if (writer == NULL) {
+        return NULL;
+    }
+
+    char *out = PyBytesWriter_GetData(writer);
     for (Py_ssize_t i = 0; i < len; i++) {
         expr_ty elem = asdl_seq_GET(strings, i);
-        PyBytes_Concat(&res, elem->v.Constant.value);
+        Py_ssize_t part = PyBytes_GET_SIZE(elem->v.Constant.value);
+        if (part > 0) {
+            memcpy(out, PyBytes_AS_STRING(elem->v.Constant.value), part);
+            out += part;
+        }
     }
-    if (!res || _PyArena_AddPyObject(arena, res) < 0) {
-        Py_XDECREF(res);
+
+    PyObject *res = PyBytesWriter_FinishWithPointer(writer, out);
```

Comment:
If PyBytesWriter_Create() is called with the exact ouput size, you can just call:

```suggestion
    PyObject *res = PyBytesWriter_Finish(writer);
```

Reply:
@vstinner 

Thank you for spotting this. It makes perfect sense, given that `PyBytesWriter_FinishWithPointer` seems to be just sugar for calculating the `size`. Do you think it makes sense to document this explicitly?

Reply:
If you think that the doc can be enhanced, go ahead.